text/0000-cargo-binary-dependencies.md

```diff
@@ -0,0 +1,140 @@
+- Building and embedding binaries for another target, such as firmware or WebAssembly. This feature would allow a versioned dependency on an appropriate crate providing the firmware or WebAssembly binary, and then embedding the binary (or a compressed or otherwise transformed version of it) into the final crate. For instance, a virtual machine could build its system firmware, or a WebAssembly runtime could build helper libraries.
+- Building and embedding a shared library for use at runtime. For instance, a binary could depend on a shared library used with [`LD_PRELOAD`](https://man7.org/linux/man-pages/man8/ld.so.8.html#ENVIRONMENT), or used in the style of the Linux kernel's [VDSO](https://man7.org/linux/man-pages/man7/vdso.7.html).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Cargo allows you to depend on binary or C ABI artifacts of another package; this is known as a "binary dependency" or "artifact dependency". For example, you can depend on the `cmake` binary in your `build.rs` like so:
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin" }
+```
+
+Cargo will build the `cmake` binary, then make it available to your `build.rs` through an environment variable:
+
+```rust
+// build.rs
+use std::{env, process::Command};
+
+fn main() {
+    let cmake_path = env::var_os("CARGO_BIN_FILE_CMAKE_cmake").expect("cmake binary");
+    let mut cmake = Command::new(cmake_path).arg("--version");
+    assert!(cmake.status().expect("cmake --version failed").success());
+}
+```
+
+You can optionally specify specific binaries to depend on using `bins`:
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin", bins = ["cmake"] }
+```
+
+If no binaries are specified, all the binaries in the package will be built and made available.
+
+You can obtain the directory containing all binaries built by the `cmake` crate with `CARGO_BIN_DIR_CMAKE`, such as to add it to `$PATH` before invoking another build system or a script.
+
+Cargo also allows depending on `cdylib` or `staticlib` artifacts. For example, you can embed a dynamic library in your binary:
+
+```rust
+// main.rs
+const MY_PRELOAD_LIB: &[u8] = include_bytes!(env!("CARGO_CDYLIB_FILE_MYPRELOAD"));
+```
+
+Note that cargo cannot help you ensure these artifacts are available at runtime for an installed version of a binary; cargo can only supply these artifacts at build time. Runtime requirements for installed crates are out of scope for this change.
+
+If you need to depend on multiple variants of a crate, such as both the binary and library of a crate, you can supply an array of strings for `type`: `type = ["bin", "lib"]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are four `type`s available:
+1. `"lib"`, the default
```

Comment:
["A Rust library will be produced. This is an ambiguous concept as to what exactly is produced because a library can manifest itself in several forms."](https://doc.rust-lang.org/reference/linkage.html).
What exactly does this mean?

Reply:
This does need clarification. "lib" here means a normal Cargo non-artifact dependency; this does not involve setting the various environment variables mentioned. A dependency with `type="lib"` is the same as a dependency without `type`.